File: constgen.py
from collections.abc import Sequence
from numpy import array, complex64, complex128, float32, float64, sqrt
from typing import Any, Callable, Tuple, TypeVar, cast


Val = TypeVar("Val")
# Vec = Tuple[Val, Val]
# print(Vec)
Vec = Sequence[Val]


def vec_ops(*, size: int, val_type: Callable[[Any], Val]):

    # Vec = Tuple.__getitem__((Val,) * size)
    # print(Vec)
    # Vec = Tuple.__getitem__((val_type,) * size)
    # print(Vec)

    class VecOps:
        @staticmethod
        def dot(a: Vec, b: Vec) -> Val:
            return sum((a[i] * b[i] for i in range(size)), val_type(0))

        @staticmethod
        def norm(a: Vec) -> Val:
            # del VecOps.dot
            return sqrt(cast(Any, VecOps.dot(a, a)))

    return VecOps


vec_ops_2f32 = vec_ops(size=2, val_type=float32)


def main():
    a = array([1.5, 2], dtype=float32)
    print(f"norm: {vec_ops_2f32.norm(a)}")


if __name__ == "__main__":
    main()

File: constgen.nim
import math
import strformat

func dot[Size: static int, Val](a, b: array[Size, Val]): Val =
  var sum: Val = 0
  for i, x in a.pairs:
    sum += x * b[i]
  sum

func norm[Size: static int, Val](a: array[Size, Val]): Val =
  dot(a, a).sqrt()

# func norm2f64(x, y: float64): float64 {.exportc.} =
#   norm([x, y])

proc main() =
  var a = [1.5, 2]
  echo fmt"norm: {norm(a)}"

main()
